src/types/UiaTextRangeBase.cpp

```diff
@@ -544,98 +544,40 @@ CATCH_RETURN();
-                                           base::ClampAdd(base::ClampSub(_end.Y, _start.Y), base::ClampedNumeric<short>(1));
-        const short lastRowInRange = _start.Y + totalRowsInRange - 1;
-
-        short currentScreenInfoRow = 0;
-        for (short i = 0; i < totalRowsInRange; ++i)
-        {
-            currentScreenInfoRow = _start.Y + i;
-            const ROW& row = buffer.GetRowByOffset(currentScreenInfoRow);
-            if (row.GetCharRow().ContainsText())
-            {
-                const size_t rowRight = row.GetCharRow().MeasureRight();
-                size_t startIndex = 0;
-                size_t endIndex = rowRight;
-                if (currentScreenInfoRow == _start.Y)
-                {
-                    startIndex = _start.X;
-                }
-
-                if (currentScreenInfoRow == _end.Y)
-                {
-                    // prevent the end from going past the last non-whitespace char in the row
-                    endIndex = std::max<size_t>(startIndex + 1, std::min(gsl::narrow_cast<size_t>(_end.X), rowRight));
-                }
+        const auto bufferSize = buffer.GetSize();
 
-                // if startIndex >= endIndex then _start is
-                // further to the right than the last
-                // non-whitespace char in the row so there
-                // wouldn't be any text to grab.
-                if (startIndex < endIndex)
-                {
-                    wstr += row.GetText().substr(startIndex, endIndex - startIndex);
-                }
-            }
+        // convert _end to be inclusive
+        auto inclusiveEnd{ _end };
+        bufferSize.DecrementInBounds(inclusiveEnd, true);
 
-            if (currentScreenInfoRow != lastRowInRange)
-            {
-                wstr += L"\r\n";
-            }
+        const auto textRects = buffer.GetTextRects(_start, inclusiveEnd);
+        const auto bufferData = buffer.GetText(/*includeCRLF*/ true,
+                                               /*trimTrailingWhitespace*/ false,
+                                               textRects);
 
-            if (getPartialText && wstr.size() > static_cast<size_t>(maxLength))
-            {
-                wstr.resize(maxLength);
-                break;
-            }
+        for (const auto& text : bufferData.text)
+        {
+            textData += text;
         }
     }
-
-    *pRetVal = SysAllocString(wstr.c_str());
-
-    // TODO GitHub #1914: Re-attach Tracing to UIA Tree
-    // tracing
-    /*ApiMsgGetText apiMsg;
-    apiMsg.Text = wstr.c_str();
-    Tracing::s_TraceUia(this, ApiCall::GetText, &apiMsg);*/
-
-#if defined(_DEBUG) && defined(UIATEXTRANGE_DEBUG_MSGS)
-    std::wstringstream ss;
-    ss << L"--------Retrieved Text Max Length(" << maxLength << L") [" << _id << L"]: " << wstr.c_str() << "\n";
-    OutputDebugString(ss.str().c_str());
-#endif
-
+    *pRetVal = SysAllocString(textData.c_str());

```

Comment:
`SysAllocString` can return null in which case you'd want to return E_OUTOFMEMORY or something.